c/meterpreter/source/extensions/stdapi/server/sys/process/memory.c

```diff
@@ -339,3 +340,359 @@ DWORD request_sys_process_memory_unlock(Remote *remote, Packet *packet)
+
+	const HANDLE open_process = GetProcAddress(kernel32_dll, "OpenProcess");
+	if (open_process == NULL) { dprintf("[MEM SEARCH] Could not get OpenProcess handle"); result = ERROR_INVALID_ADDRESS; goto done; }
+
+	const HANDLE close_handle = GetProcAddress(kernel32_dll, "CloseHandle");
+	if (close_handle == NULL) { dprintf("[MEM SEARCH] Could not get CloseHandle handle"); result = ERROR_INVALID_ADDRESS; goto done; }
+
+	const HANDLE nt_read_virtual_memory = GetProcAddress(ntdll_dll, "NtReadVirtualMemory");
+	if (nt_read_virtual_memory == NULL) { dprintf("[MEM SEARCH] Could not get NtReadVirtualMemory handle"); result = ERROR_INVALID_ADDRESS; goto done; }
+
+	const OPENPROCESS OpenProcess = (OPENPROCESS)open_process;
+	const CLOSEHANDLE CloseHandle = (CLOSEHANDLE)close_handle;
+	const VIRTUALQUERYEX VirtualQueryEx = (VIRTUALQUERYEX)virtual_query_ex;
+	const NTREADVIRTUALMEMORY NtReadVirtualMemory = (NTREADVIRTUALMEMORY)nt_read_virtual_memory;
+
+	const DWORD process_vm_read = 0x0010;
+	const DWORD process_query_information = 0x0400;
+	const DWORD wanted_process_perms = process_vm_read | process_query_information;
+
+	dprintf("[MEM SEARCH] Opening process");
+	const HANDLE process_handle = OpenProcess(wanted_process_perms, FALSE, pid);
+	if (process_handle == NULL) { dprintf("[MEM SEARCH] Could not get process handle"); result = ERROR_INVALID_HANDLE; goto done; }
+
+	MEMORY_BASIC_INFORMATION mem = { 0 };
+	const size_t megabytes_64 = 64 * 1024 * 1024;
+
+	dprintf("[MEM SEARCH] Allocating buffer for storing process memory");
+	buffer = (char*)malloc(megabytes_64);
+	if (buffer == NULL) { dprintf("[MEM SEARCH] Could not allocate memory buffer"); result = ERROR_OUTOFMEMORY; goto done; }
+
+	// The maximum length of data that we can read into a buffer at a time from a memory region.
+	const size_t current_max_size = megabytes_64;
+
+	for (PVOID current_ptr = NULL; VirtualQueryEx(process_handle, current_ptr, &mem, sizeof(mem)); (size_t)current_ptr += mem.RegionSize)
+	{
+		if (!can_read_memory(mem.Protect)) { continue; }
+
+		size_t memory_region_offset = 0;
+		// Note: This currently does not support regex'ing over multiple memory regions.
+		// e.g.
+		// regex = "my_password.*";
+		// | ....my_pas | sword.... |
+		while (mem.RegionSize > memory_region_offset)
+		{
+			const size_t leftover_bytes = mem.RegionSize - memory_region_offset;
+			const size_t bytes_to_read = min(leftover_bytes, current_max_size);
+			dprintf("[MEM SEARCH] Leftover Bytes count: %llu", leftover_bytes);
+			dprintf("[MEM SEARCH] Bytes to read: %llu", bytes_to_read);
+			size_t bytes_read = 0;
+
+			const size_t read_address = (size_t)mem.BaseAddress + memory_region_offset;
+			// Note: This will read up to a maximum of bytes_to_read OR to the end of the memory region if the end of it has been reached.
+			NtReadVirtualMemory(process_handle, (LPCVOID)read_address, buffer, bytes_to_read, &bytes_read);

```

Comment:
Do we need to check the return values here? 👀 

https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-ntallocatevirtualmemory#return-value

Reply:
That's for allocating virtual memory, but I did add in a check for `ERROR_SUCCESS` 🚀